Assets/Scripts/Characters/Player.cs

```diff
@@ -35,66 +49,138 @@ private void Awake()
                     _deathAnimationTime = clip.length;
                 }
             }
+        }
 
-            PlayerEnergy.EnergyExhausted += HandleEnergyExhausted;
-            PlayerEnergy.EnergyReset += HandleEnergyReset;
+        public int GetHealth()
+        {
+            return _currentHealth;
         }
 
-        private void OnDestroy()
+        public bool IsAlive()
         {
-            PlayerDespawned?.Invoke(this);
-            PlayerEnergy.EnergyExhausted -= HandleEnergyExhausted;
-            PlayerEnergy.EnergyReset -= HandleEnergyReset;
+            return _currentHealth > 0;
         }
 
-        private void HandleEnergyExhausted()
+        #region Server
+
+        public override void OnStartServer()
         {
-            if (_playerMovement)
-            {
-                _playerMovement.gameObject.SetActive(false);
-            }
-            if (_playerCombat)
+            PlayerEnergy.ServerEnergyExhausted += ServerHandleEnergyExhausted;
+            PlayerEnergy.ServerEnergyReset += ServerHandleEnergyReset;
+
+            ServerOnPlayerSpawned?.Invoke(this);
+        }
+
+        public override void OnStopServer()
+        {
+            PlayerEnergy.ServerEnergyExhausted -= ServerHandleEnergyExhausted;
+            PlayerEnergy.ServerEnergyReset -= ServerHandleEnergyReset;
+
+            ServerOnPlayerDespawned?.Invoke(this);
+        }
+
+        private void ServerHandleEnergyExhausted(int connectionId)
+        {
+            if (connectionToClient.connectionId != connectionId)
             {
-                _playerCombat.gameObject.SetActive(false);
+                return;
             }
+
+            RpcHandleEnergyExhaustion();
         }
 
-        private void HandleEnergyReset()
+        private void ServerHandleEnergyReset(int connectionId)
         {
-            if (_playerMovement)
+            if (connectionToClient.connectionId != connectionId)
             {
-                _playerMovement.gameObject.SetActive(true);
+                return;
             }
-            if (_playerCombat)
+            RpcHandleEnergyReset();
+        }
+
+        [Server]
+        public void TakeDamage(int damage)
+        {
+            _currentHealth -= damage;
+            Debug.Log($"Player maxHealth is {maxHealth}");
+            if (!IsAlive())
             {
-                _playerCombat.gameObject.SetActive(true);
+                RpcOnDeath();
             }
         }
 
-        private void Start()
+        IEnumerator DestroyWithAnimation()
```

Comment:
This is basically a clean way handle logic that takes multiple frames

https://docs.unity3d.com/Manual/Coroutines.html
https://docs.unity3d.com/ScriptReference/MonoBehaviour.StartCoroutine.html